File: valid_palindrome.cxx
class Solution {
public:
    /*bool isPalindrome(string s) {
        std::erase_if(s,[](char const &c) {return !std::isalnum(c);});
        std::transform(s.begin(),s.end(),s.begin(),[](unsigned char c){return std::tolower(c);});
        auto t = s;
        std::reverse(t.begin(),t.end());
        return s == t;
    }*/
    
    bool isPalindrome(string s) {
        bool ret=true;
        for(auto i=s.begin(),j=s.end()--; i<j; ++i,--j) {
            for(;i<j && !std::isalnum(*i); ++i); //shorthand while loop
            for(;i<j && !std::isalnum(*j); --j); //shorthand while loop
            if(std::tolower(*i)!=std::tolower(*j)) {
                ret=false;
                break;
            }    
        }
        return ret;
    }
};

File: valid_palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        s = [i for i in s.lower() if i.isalnum()]
        return all(s[i] == s[~i] for i in range(len(s) // 2))
    
    #def isPalindrome(self, s: str) -> bool:
    #    s = [i for i in s.lower() if i.isalnum()]
    #    return s == s[::-1]


# remove non alphanumeric -> lowercase -> reverse -> equate
# start from beginning and end, skip non alphanumeric, check char by char tolower till we hit middle
